File: cpp/091_min_cost/min_cost.cpp
/*
 * @Date: 2022-06-25
 * @LastEditors: Mengsen Wang mengsen_wang@163.com
 * @LastEditTime: 2022-06-25
 * @FilePath: /algorithm/091_min_cost/min_cost.cpp
 */

#include <cassert>
#include <vector>

using namespace std;

class Solution {
 public:
  int minCost(vector<vector<int>> costs) {
    int n = costs.size();
    vector<int> dp(3);
    for (int j = 0; j < 3; j++) {
      dp[j] = costs[0][j];
    }
    for (int i = 1; i < n; i++) {
      vector<int> dpNew(3);
      for (int j = 0; j < 3; j++) {
        dpNew[j] = min(dp[(j + 1) % 3], dp[(j + 2) % 3]) + costs[i][j];
      }
      dp = dpNew;
    }
    return *min_element(dp.begin(), dp.end());
  }
};

int main() {
  assert(Solution().minCost(vector<vector<int>>{{17, 2, 17}, {16, 16, 5}, {14, 3, 19}}) == 10);
  assert(Solution().minCost(vector<vector<int>>{{7, 6, 2}}) == 2);
}

File: golang/091_min_cost/min_cost.go
/*
 * @Date: 2022-06-25
 * @LastEditors: Mengsen Wang mengsen_wang@163.com
 * @LastEditTime: 2022-06-25
 * @FilePath: /algorithm/091_min_cost/min_cost.go
 */

package main

func minCost(costs [][]int) int {
	min := func(a, b int) int {
		if a > b {
			return b
		}
		return a
	}
	dp := costs[0]
	for _, cost := range costs[1:] {
		dpNew := make([]int, 3)
		for j, c := range cost {
			dpNew[j] = min(dp[(j+1)%3], dp[(j+2)%3]) + c
		}
		dp = dpNew
	}
	return min(min(dp[0], dp[1]), dp[2])
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}
	assert(minCost([][]int{{17, 2, 17}, {16, 16, 5}, {14, 3, 19}}) == 10)
	assert(minCost([][]int{{7, 6, 2}}) == 2)
}
